### ISSUE TITLE
Sanity check $entity->original in filter_entity_update in filter module.

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#2624]
I am having an error report that I think may be due to problem with some code in filter/filter.module. In function filter_entity_update which implements hook_entity_update at line 1615 there is
``` 
if (!empty($entity->original) && .......
```
then at line 1626 there is the else condition
```  
else {
    $original_fids_by_field = _filter_get_file_ids_by_field($entity->original);
``` 
which is failing because in some circumstances $entity->original does not exist and _filter_get_file_ids_by_field fails.

I have found this when developing a proposed contributed module 'trigger-actions' and it manifests itself when in testing I set an action to change a node's status to 'published' even if saved as unpublished. It has had me puzzled for several days!

---
~PR by @Graham-72: https://github.com/backdrop/backdrop/pull/1830~
PR by @Graham-72: https://github.com/backdrop/backdrop/pull/1851

**Comment by Graham-72** [comment]
More information:
I hit this problem when attempting to use function node_publish_action for an existing node i.e. an update.
It seems the value $entity->original is being removed somewhere.
I have traced this to the following code in node/node.entity.inc at line 295 in public function save
```
      // Ignore slave server temporarily.
      db_ignore_slave();
      unset($entity->original);
```
Why is $entity->original being unset here? As I do not know the purpose of this code I cannot produce a suitable PR. Please can someone take a look at this for me.

**Comment by docwilmot** [comment]
@Graham-72 AFAICT that `unset($entity->original)` runs *after* the update hook is called.
```
  public function save(EntityInterface $entity) {
        $op = 'update';
      ...
      $this->invokeHook($op, $entity);
      ...
      unset($entity->original);

      return $return;
    }
```

So when `filter_entity_update()` is called by that `invokeHook()` the `$entity->original` hasn't yet been unset.

BUT if another module somewhere is unsetting `$entity->original` in its update hook, and it runs before Filter, that might be the problem. Can you search for `unset($entity->original)` in your installed contribs?

**Comment by docwilmot** [comment]
And check that `$entity->original` exists *before* the `invokeHook()`?

**Comment by Graham-72** [comment]
@docwilmot Thanks for your suggestions. I have been trying to track what is happening here. 
When I edit and save a node in the normal way it invokes two hooks: presave and update.

When I set trigger_actions to, for example, 'promote content' when saving a post, action 'node_promote_action' is triggered and then a total of 4 hooks are invoked, two presaves, and two updates. 

`node.entity.inc` is unsetting `$entity->original` in the first of these update hooks, after `filter_entity_update()` is called (as you say) and the update is successfully applied to the node.

But for some reason I do not understand the update hook is called again, I think by `protected function invokeHook()` at the line `parent::invokeHook($hook, $node);` This time `$entity->original` has been unset and so `filter_entity_update()` fails.

I think there must be something wrong in the logic of this sequence.

**Comment by docwilmot** [comment]
I see your point. And the problem is actually simple:

D7 Action callbacks only set properties on the entity, and left the CRUD actions to the entity controller. So the `node_unpublish_action` callback looks like this:
```
function node_unpublish_action($node, $context = array()) {
  $node->status = NODE_NOT_PUBLISHED;
  watchdog('action', 'Set @type %title to unpublished.', array('@type' => node_type_get_name($node), '%title' => $node->title));
}
```
So when trigger/actions ran on a node save, Actions just set the status of the node being saved, thats it. Drupal did the saving. 

But Backdrop added a save call to the callbacks:
```
function node_unpublish_action(Node $node, &$context) {
  $node->status = NODE_NOT_PUBLISHED;
  $node->save();
}
```
So now when a node is saving, your Action runs in the update hook, calls `save()`, then the node continues with the save it started off with in the first place.

**Comment by Graham-72** [comment]
@docwilmot thanks for the explanation. So Backdrop's actions 'node_unpublish_action' etc. are unusable? I.e. this is a bug that needs fixing somehow?

**Comment by docwilmot** [comment]
@Graham-72 I cannot think of a way to make this work without duplicating core action callbacks.

**Comment by Graham-72** [comment]
@docwilmot All seems OK if I rearrange the test in filter.module as in this PR.

**Comment by docwilmot** [comment]
Not so sure @Graham-72.
For one this seems off; was there a typo here?
```
if (!empty($entity->original)){
    $vid_key && $entity->$vid_key != $entity->original->$vid_key) {
```

Secondly though, even if this PR were merged, we still have the unacceptable issue of `save()` et al running twice.

**Comment by Graham-72** [comment]
@docwilmot Thanks for spotting my careless typo. I have amended the PR.

Are you thinking that the running of `save()` twice is the responsibility of trigger_actions because I am thinking it is a Backdrop Core problem that will be experienced by anything trying to use one of these core actions?

**Comment by docwilmot** [comment]
@Graham-72 thinking of possible solutions and coming up short. I think it was reasonable for Backdrop to make those changes, seeing as Trigger was removed, but now there's no way to correct the fact that Action callbacks cause a save. So really, I think this module will have to build an entire new mechanism (a copy of D7 Actions) for using triggers rather than using Backdrop's core Actions.

**Comment by Graham-72** [comment]
@docwilmot yes, I think you are right. I will see what can be done by providing new actions in the trigger_actions module that avoid the double save.
Though I do think that the Backdrop code in filter.module needs some attention to deal with the situation where $entity->original is empty, and not just fail.

**Comment by docwilmot** [comment]
@Graham-72 the core code already accounts for this; there is a check at the start of the `save()`:
```
      // Load the stored entity, if any.
      if (!$entity->isNew() && !isset($entity->original)) {
        $entity->original = entity_load_unchanged($this->entityType, $entity->id());
      }
      ...
      $this->invokeHook($op, $entity);
      ...
      unset($entity->original);
```

**Comment by Graham-72** [comment]
I have added new actions to the trigger_actions contrib module that avoid this problem, but this does not prevent someone falling into what I see as this software trap, which needs to be fixed.

**Comment by Dinornis** [comment]
Missing link to PR?

**Comment by klonos** [comment]
Added the link to the PR in the issue summary and also added appropriate tags. I think that this issue could use some feedback from @quicksketch and/or @serundeputy

**Comment by quicksketch** [comment]
This looks good to me. Sorry about the trouble with actions @Graham-72. This has been a stumbling block for upgrading action-related modules for sure. We made some compromises when Trigger was taken out and Views Bulk Operations was put in, refocusing the action system to handle that situation as its primary use-case instead of Trigger.

There was a single missing space in the PR, so I rolled a new one that corrects this minor issue. https://github.com/backdrop/backdrop/pull/1851
